test/Test/Database/LSMTree/StateMachine.hs

```diff
@@ -465,97 +469,98 @@ createSystemTempDirectory prefix = do
-    handler' e                            = Just (Model.ErrOther (displayException e))
-
-commitActionRegistryErrorHandler :: Monad m => Handler m (Maybe Model.Err)
-commitActionRegistryErrorHandler = Handler $ \(e :: CommitActionRegistryError) ->
-  pure (classifyException (toException e))
-
-abortActionRegistryErrorHandler :: Monad m => Handler m (Maybe Model.Err)
-abortActionRegistryErrorHandler = Handler $ \(e :: AbortActionRegistryError) ->
-  pure (classifyException (toException e))
-
--- | Some exceptions contain other exceptions, which we classify recursively.
-classifyException :: SomeException -> Maybe Model.Err
-classifyException e =
-  Just . fromMaybe (Model.ErrOther (displayException e)) $ classifyException' e
-
--- | When classifying exceptions recursively, we prefer 'Model.ErrDiskFault'
---   and 'Model.ErrSnapshotCorrupted' as explanations over 'Model.ErrOther'.
-classifyException' :: SomeException -> Maybe Model.Err
-classifyException' e
-  | Just (CommitActionRegistryError es) <- fromException e
-  = classifyExceptions' es
-  | Just (AbortActionRegistryError reason es) <- fromException e
-  = (classifyException' =<< getReasonExitCaseException reason) <|> classifyExceptions' es
-  | Just (e' :: ActionError) <- fromException e
-  = classifyException' (getActionError e')
-  | Just FsError{} <- fromException e
-  = Just (Model.ErrDiskFault (displayException e))
-  | Just FileFormatError{} <- fromException e
-  = Just (Model.ErrSnapshotCorrupted (displayException e))
-  | Just ChecksumsFileFormatError{} <- fromException e
-  = Just (Model.ErrSnapshotCorrupted (displayException e))
-  | Just ChecksumError{} <- fromException e
-  = Just (Model.ErrSnapshotCorrupted (displayException e))
-  | otherwise
-  = Nothing
-
-classifyExceptions' :: (Foldable t, Exception e) => t e -> Maybe Model.Err
-classifyExceptions' = getFirst . foldMap (First . classifyException' . toException)
-
-fsErrorHandler :: Monad m => Handler m (Maybe Model.Err)
-fsErrorHandler = Handler $ pure . handler'
-  where
-    handler' :: FsError -> Maybe Model.Err
-    handler' e = Just (Model.ErrDiskFault (displayException e))
+handleCommitActionRegistryError :: CommitActionRegistryError -> Model.Err
+handleCommitActionRegistryError = \case
+  CommitActionRegistryError actionErrors ->
+    Model.ErrCommitActionRegistry $
+      handleSomeException <$> (getActionError <$> actionErrors)
 
-fileFormatErrorHandler :: Monad m => Handler m (Maybe Model.Err)
-fileFormatErrorHandler = Handler $ pure . handler'
-  where
-    handler' :: FileFormatError -> Maybe Model.Err
-    handler' e = Just (Model.ErrSnapshotCorrupted (displayException e))
+handleAbortActionRegistryError :: AbortActionRegistryError -> Model.Err
+handleAbortActionRegistryError = \case
+  AbortActionRegistryError abortReason actionErrors ->
+    Model.ErrAbortActionRegistry $
+      handleSomeException <$> maybe id NE.cons (getReasonExitCaseException abortReason) (getActionError <$> actionErrors)
```

Comment:
I think `ErrAbortActionRegistry` should keep the exit case reason separate. This is going to be a type that users can see in counter examples, so it's useful to see the distinction  between the exit case reason and any exceptions that happen while commiting / aborting the action registry